e2e/auth.setup.ts

```diff
@@ -0,0 +1,110 @@
+import {
+  USER_NAME_LOCALSTORAGE_KEY,
+  createTestUser,
+  deleteTestUser,
+  getAccessTokenFromCookie,
+  getAdminUserCredentials,
+  loginTestUser,
+} from './helpers/testUser'
+import { TestUser } from './helpers/types'
+
+let testUserId: string | undefined
+let adminAccessToken: string | undefined
+let adminPage: Page
+let userPage: Page
+
+setup.describe('Setup', () => {
+  setup.beforeAll(async ({ browser }) => {
+    adminPage = await browser.newPage()
+    userPage = await browser.newPage()
+  })
+
+  setup('authenticate users', async () => {
+    setup.slow()
+
+    await setup.step('authenticate admin', async () => {
+      const { adminUserName, adminUserPassword } = getAdminUserCredentials()
+
+      await loginTestUser(adminPage, adminUserName!, adminUserPassword!)
+      await adminPage.context().storageState({ path: ADMIN_STORAGE_STATE })
+
+      adminAccessToken = await getAccessTokenFromCookie(adminPage)
+    })
+
+    const { testUserName, testUserPassword } = await setup.step(
+      'create test user',
+      async () => {
+        const testUserName = 'swc-e2e-' + uuidv4()
+        const testUserPassword = 'password-' + uuidv4()
+        /* 
+        const testUserName = 'swc-e2e-test'
+        const testUserPassword = process.env.USER_PASSWORD!
+        expect(testUserPassword).not.toBeUndefined() 
+        */
+
+        const testUserEmail = `${testUserName}@test.com`
+        const testUserIsValidated = false
+
+        const testUser: TestUser = {
+          username: testUserName,
+          email: testUserEmail,
+          password: testUserPassword,
+          tou: true,
+          validatedUser: testUserIsValidated,
+        }
+
+        testUserId = await createTestUser(
+          getEndpoint(),
+          testUser,
+          adminAccessToken,
+        )
+        expect(testUserId).not.toBeUndefined()
+
+        await setLocalStorage(
+          userPage,
+          USER_NAME_LOCALSTORAGE_KEY,
+          testUserName,
+        )
+        await userPage.context().storageState({ path: USER_STORAGE_STATE })
+
+        return { testUserName, testUserPassword }
+      },
+    )
+
+    await setup.step('authenticate test user', async () => {
+      await loginTestUser(userPage, testUserName, testUserPassword)
+      await userPage.context().storageState({ path: USER_STORAGE_STATE })
+    })
+  })
+
+  setup.afterAll(async ({}, { expectedStatus, status }) => {
+    // clean up pages
+    await adminPage.close()
+    await userPage.close()
+
+    // delete test user if setup wasn't successful:
+    // ...cleanup isn't run until after last retry
+    // ...so we delete here, so we can retry setup without leaving behind test users
+    if (status !== expectedStatus && testUserId !== undefined) {
+      const result = await deleteTestUser(
+        getEndpoint(),
+        testUserId!,
+        adminAccessToken,
+      )
+      expect(result).toEqual(testUserId)
+    }
+  })
```

Comment:
We create the test user in `auth.setup` and delete the test user in `auth.cleanup`. However, if `auth.setup` is retried, then `auth.cleanup` won't be run until after the last retry. Delete the test user here if the test failed or was skipped, so that setup can be retried without leaving behind test users.

Reply:
Great work handling this edge case